vnext/Microsoft.ReactNative/Modules/DeviceInfoModule.cpp

```diff
@@ -45,15 +49,40 @@ void DeviceInfoHolder::InitDeviceInfoHolder(
           });
     } else {
       assert(react::uwp::IsXamlIsland());
-      // TODO: WinUI 3 Islands - set up a listener for window size changed
+      // This is either a WinUI 3 island or a system XAML island
+      // system XAML islands have a CoreWindow so we want to use the GetForCurrentView APIs
+      // For WinUI 3 islands we require the app to forward window messages as ReactNotifications
     }
 
-    deviceInfoHolder->m_dpiChangedRevoker = displayInfo.DpiChanged(
-        winrt::auto_revoke, [weakHolder = std::weak_ptr(deviceInfoHolder)](const auto &, const auto &) {
-          if (auto strongHolder = weakHolder.lock()) {
-            strongHolder->updateDeviceInfo();
-          }
-        });
+    if (!react::uwp::IsWinUI3Island()) {
+      // UWP or system XAML island
+      auto const &displayInfo = winrt::Windows::Graphics::Display::DisplayInformation::GetForCurrentView();
+
+      deviceInfoHolder->m_dpiChangedRevoker = displayInfo.DpiChanged(
+          winrt::auto_revoke, [weakHolder = std::weak_ptr(deviceInfoHolder)](const auto &, const auto &) {
+            if (auto strongHolder = weakHolder.lock()) {
+              strongHolder->updateDeviceInfo();
+            }
+          });
+    } else if (auto hwnd = (HWND)XamlUIService::GetIslandWindow(deviceInfoHolder->m_context->Properties())) {
+      deviceInfoHolder->m_wmSubscription = SubscribeToWindowMessage(
+          context.Notifications(),
+          WM_WINDOWPOSCHANGED,
+          [weakHolder = std::weak_ptr(deviceInfoHolder)](HWND hwnd, const DesktopWindowMessage &dwm) {
+            if (auto strongHolder = weakHolder.lock()) {
+              const auto pos = reinterpret_cast<WINDOWPOS *>(dwm.lParam());
+              const auto newWidth = (float)(pos->cx);
```

Comment:
>(float) [](start = 36, length = 7)

Generally in modern C++ we must avoid use of the C-style cast. In this case it is better to use the static_cast.